core/embed/rust/src/protobuf/decode.rs

```diff
@@ -289,17 +289,125 @@ impl<'a> InputStream<'a> {
         Ok(val)
     }
 
-    pub fn read_uvarint(&mut self) -> Result<u64, Error> {
-        let mut uint = 0;
+    pub fn read_uvarint_u32(&mut self) -> Result<u32, Error> {
+        let mut uint: u32 = 0;
         let mut shift = 0;
-        loop {
-            let byte = self.read_byte()?;
-            uint += (byte as u64 & 0x7F) << shift;
+        for byte_count in 1..=5 {
+            let byte: u32 = self.read_byte()? as u32;
+            let continue_decoding = byte & 0x80 != 0;
+            let payload = byte & 0x7F;
+            if byte_count == 5 {
+                // first 4 bytes are accepted as is (4 x 7 bits = 28 bits),
+                // then the 5th byte is only accepted it it is the final byte
+                // and contains no more than 4 useful bits! (4 x 7 bits + 4 bits = 32 bits)!
+                if continue_decoding || payload > 0xF {
+                    return Err(error::overflow());
+                }
+            }
+            uint += payload << shift;
+            shift += 7;
+            if !continue_decoding {
+                break;
+            }
+        }
+        Ok(uint)
```

Comment:
nit: maybe you could get a tiny bit cleaner code by reversing this:
"if !continue_decoding { return Ok(uint) }" + in the failing cases break the loop to an Err(Overflow) result
(that lets you not keep explicit track of `continue_decoding`, and instead say "if byte_count == 5 && payload > 0xF { break }")